packages/protocol/contracts/layer1/fork/ForkManager.sol

```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import "@openzeppelin/contracts/proxy/utils/UUPSUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/Ownable2StepUpgradeable.sol";
+
+/// @title ForkManager
+/// @custom:security-contact security@taiko.xyz
+/// @notice This contract serves as a base contract for managing up to two forks within the Taiko
+/// protocol. By default, all function calls are routed to the newFork address.
+/// Sub-contracts should override the shouldRouteToOldFork function to route specific function calls
+/// to the old fork address.
+/// These sub-contracts should be placed between a proxy and the actual fork implementations. When
+/// calling upgradeTo, the proxy should always upgrade to a new ForkManager implementation, not an
+/// actual fork implementation.
+/// It is strongly advised to name functions differently for the same functionality across the two
+/// forks, as it is not possible to route the same function to two different forks.
+///
+///                         +--> newFork
+/// PROXY -> FORK_MANAGER --|
+///                         +--> oldFork
+
+contract ForkManager is UUPSUpgradeable, Ownable2StepUpgradeable {
```

Comment:
An instance of this `ForkManager` (even behind proxy) is not really upgradeable, or at least not the 2 main variables (oldFork, newFork). 
Since you placed here a constructor and not `init`, i suppose it is intentional from your side, but in this case we dont need the `Upgradeable` pattern. (Or if we do want to do that, just remove the immutability of the 2 fields).

Reply:
`oldFork` and `currFork` are not saved in storage as in a normal proxy, they are complied into the deployed code by the constructor and immutable variables..